README.rst
JCchess - John Cheetham - http://www.johncheetham.com/projects/jcchess
 
Description
-----------
JCchess is a program to play chess against UCI chess engines.

Requirements
------------
python3 python-cairo python-gobject gtk3

Usage
-----
Enter 'python3 run.py' to start the program.

Adding a UCI engine
-------------------
To add an engine to play against click on Options, engines then click
the 'Add' button to add a new engine. Navigate to the engine executable
and add it. Then click the OK button.

Next do Options, players and set the black player to the new engine.

Click on the 'tick' button at the top of the screen to start the game 
and then move a white piece.

For debugging start with -v (all debugging messages) or -vuci (uci
messages).

Acknowledgements
----------------
The initial version was based on gshogi.
see http://www.johncheetham.com/projects/gshogi

python-chess is used.
see https://github.com/niklasf/python-chess

Piece Images by DG-RA
see https://openclipart.org/user-detail/DG-RA


ChangeLog

2017-04-01 John Cheetham
    * Initial commit of jcchess based on gshogi code


Minor modifications for 0.5.1
----------------------------------
2017-03-01 Bernd Wille
    * Bugfix hanging 3rd party engines during ponder
    * Bugfix in engine configuration (floats in utils.py 771)

    * works well with Spear 2009 v1.5, GPSShogi r2837, Laramie V3, Lesserkai 1.4.1. All Blunder-engines fail sending
      characters I have not been able to convert to utf-8 in a safe way
    * be sure to set the number of threads to number of cores of your machine -1. Otherwise they will block
      the communication in the usi-interface and frequent blocks will result
    * Command-line -m switches to a version with moves and comments in main window. 
    * Command-line -h shows header of loaded files. -mh shows both
    * resigns and checkmates written into comments by the program -- at least in many cases
   
2017-2-04  Bernd Wille
	* Bugfix gshogi.py line 765: empty comment-list when starting new game

2017-0-18 Bernd Wille
	* Remember last used path for loaded games
	* Show filename in Window-Titlebar
	* take filename from command-line and enable starting games by double-click, use gshogirun-script (please edit paths).


Changes for 0.5.0
-----------------

    * Add i18n support

    * Add new pieces and colour scheme

    * Run on Windows

    * Use cairo to add pieces to board

    * Port to Python 3 / GTK 3 (PyGI)

    * Use setuptools instead of distutils

    * Bugfixes


Changes for 0.4.5
-----------------
2012-08-28  John Cheetham  developer@johncheetham.com

    * Add new gamelist window and support for reading multi-game PSN files.

    * Add a new comments window so you can add comments to the game.

    * Allow loading of custom piece images

    * Apply Fixes to make pondering work

    * Add ponder move and current move to engine output window

    * Remove the none.jpg for empty squares. Empty pixbuf now created in code.

    * Add score and change time format in engine output.


Changes for 0.4.4
-----------------
2012-07-23  John Cheetham  developer@johncheetham.com

    * Add copy/paste of position/game to clipboard

    * Add Edit Board function

    * Add move highlighting

    * Add co-ordinates for board squares

    * Improve the look of the komadai

    * Move code for drag and drop, load/save to separate modules

    * Convert main window to glade - some widgets only

    * Fix bug where side to move is shown wrong after clicking
      a move in the movelist

    * Fix bug - Fix AttributeError in gshogi.py and Drag_and_Drop.py

    * Fix bug - connect delete-event signal in engine-output.py

    * Fix bug - invalid textbuf iterator

    * Delete duplicated function (get_startpos) in gshogi.py

    * Fix bug in processing of position SFEN line


Changes for 0.4.3
-----------------
2012-06-29  John Cheetham  developer@johncheetham.com

    * Add a menu item to allow customisable colour schemes

    * Add an engine debug window for USI engines

    * Add a move list window

    * Add an engine ourput window for USI engines

    * Add a common engine settings dialog for ponder/hash size

    * Now Uses glade for any new menus


Changes for 0.4.2
-----------------
2011-04-10  John Cheetham  developer@johncheetham.com

    * Add Game Navigation buttons to toolbar

    * Allow Drag and Drop of pieces


Changes for 0.4.1
-----------------
2011-04-03  John Cheetham  developer@johncheetham.com

    * Make board resizing quicker after resizing the main window


Changes for 0.4
---------------
2011-03-23  John Cheetham  developer@johncheetham.com

    * Add Menu option to play handicap games

    * Add new USI time controls - classical, incremental, depth, nodes,
      fixed time per move and infinite

    * Add support for loading/saving games in the PSN format

    * suppress debugging console messages by default. Print them only
      if the '-v' (verbose) command line switch is used.


Changes for 0.3
---------------
2010-10-03  John Cheetham  developer@johncheetham.com

    * New set player function to allow playing engine vs engine as
      well as human vs engine and human vs human.

    * Change Time Control to use available time and byoyomi

    * Add toolbar showing players, clocks and stop/go buttons

    * Remember the setting of Eastern/Western pieces, time controls
      and engines

    * set gameover to false when loading game

    * fix abends in engine manager when you press delete/rename and
      no engine was selected.

    * Retract Move has been renamed as Undo move and undoes 1 move
      and not a pair of moves.

    * A redo move function has been added.


Changes for 0.2.1
-----------------
2010-08-28  John Cheetham  developer@johncheetham.com

    * Fix bug where some controls are unresponsive after black mates

    * Improve combo boxes on USI configure screen
      remember setting from previously
      increase field width

    * Cosmetic changes to configure dialogue boxes

    * improve USI engine startup time


Changes for 0.2
---------------
2010-08-20  John Cheetham  developer@johncheetham.com

    * Add USI support

    * Add Move Now option


Changes for 0.1.1
-----------------
2010-07-29  John Cheetham  developer@johncheetham.com

    * Make window resizable

    * Remove warning messages that appear when compiling in Ubuntu 9.04

    * fix eastern piece set transparent pngs

    * suppress beeps after moves (you can turn them on in constants.py)


Inital Version 0.1
------------------
2010-07-14  John Cheetham  developer@johncheetham.com

    * Release the initial version (0.1)


jcchess/pieces.py
#
#   pieces.py
#
#   This file is part of jcchess
#
#   jcchess is free software: you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation, either version 3 of the License, or
#   (at your option) any later version.
#
#   jcchess is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with jcchess.  If not, see <http://www.gnu.org/licenses/>.
#

from gi.repository import Gdk
from gi.repository import GdkPixbuf
import os
import traceback
from . import chess


class Pieces:

    def __init__(self):
        self.pieceset = "jcchess"     # jcchess, eastern, western or custom
        self.custom_piece_pixbuf = None
        self.custom_piece_path = None

        # create pixbuf for empty square
        self.pb_empty = GdkPixbuf.Pixbuf.new(
            GdkPixbuf.Colorspace.RGB, True, 8, 64, 64)
        self.pb_empty.fill(0xffffff00)  # fill with transparent white

    # called from jcchess.py
    def load_pieces(self, prefix):
        #  0 - unoccupied
        #  1 - black pawn
        #  2 - black lance
        #  3 - black knight
        #  4 - black silver general
        #  5 - black gold general
        #  6 - black bishop
        #  7 - black rook
        #  8 - black king
        #  9 - promoted black pawn
        # 10 - promoted black lance
        # 11 - promoted black knight
        # 12 - promoted black silver general
        # 13 - promoted black bishop
        # 14 - promoted black rook
        # 15 - white pawn
        # 16 - white lance
        # 17 - white knight
        # 18 - white silver general
        # 19 - white gold general
        # 20 - white bishop
        # 21 - white rook
        # 22 - white king
        # 23 - promoted white pawn
        # 24 - promoted white lance
        # 25 - promoted white knight
        # 26 - promoted white silver general
        # 27 - promoted white bishop
        # 28 - promoted white rook

        # load builtin jcchess pieces as pixbufs
        path = os.path.join(prefix, "images", "jcchess")
        self.jcchess_piece_pixbuf, errmsg = self.load_pixbufs(path)
        if errmsg is not None:
            print("Error loading jcchess pieces:",errmsg)
            return 1

        # load builtin gnushogi eastern pieces as pixbufs
        #path = os.path.join(prefix, "images", "eastern")
        #self.eastern_piece_pixbuf, errmsg = self.load_pixbufs(path)
        #if errmsg is not None:
        #    print("Error loading eastern pieces:",errmsg)
        #    return 1

        # load builtin gnushogi western pieces as pixbufs
        #path = os.path.join(prefix, "images", "western")
        #self.western_piece_pixbuf, errmsg = self.load_pixbufs(path)
        #if errmsg is not None:
        #    print("Error loading western pieces:",errmsg)
        #    return 1

        # load custom piece pixbufs if any
        if self.custom_piece_path is not None:
            self.custom_piece_pixbuf, errmsg =  self.load_pixbufs(
                self.custom_piece_path)
            # if pieces set to custom and they cannot be loaded then switch to
            # eastern
            if self.custom_piece_pixbuf is None:
                if errmsg is not None:
                    print(errmsg)
                if self.pieceset == "custom":
                    self.pieceset = "jcchess"

    def custom_pieces_loaded(self):
        if self.custom_piece_pixbuf is not None:
            return True
        else:
            return False

    # method to load images of pieces
    def load_pixbufs(self, pieces_dir):
        images = [
            "pawn", "knight",
            "bishop", "rook", "king", "queen",
            ]        

        piece_pixbuf = []
        # first pixbuf in list is empty square
        piece_pixbuf.append(self.pb_empty.copy())

        # get file extension (png or svg)
        image = images[0] + "B"
        if os.path.isfile(os.path.join(pieces_dir, image + ".png")):
            extension = ".png"
        elif os.path.isfile(os.path.join(pieces_dir, image + ".svg")):
            extension = ".svg"
        else:
            return None, "Error loading pieces\n\nFile not " \
                        "found:pawnB.png or pawnB.svg"

        # Load black pieces
        for image in images:
            # custom pieces
            image = image + "B"+ extension
            path = os.path.join(pieces_dir, image)
            if not os.path.isfile(path):
                errmsg = "Error loading pieces\nFile not found:" + image
                return None, errmsg
            pb = GdkPixbuf.Pixbuf.new_from_file(path)
            piece_pixbuf.append(pb)

        # Load white pieces
        for image in images:
            # change filename from pawnB to pawnW etc
            image_white = image + "W" + extension

            # if user has provided an image for white then use it
            path = os.path.join(pieces_dir, image_white)
            if os.path.isfile(path):
                pb = GdkPixbuf.Pixbuf.new_from_file(path)
                piece_pixbuf.append(pb)
                continue

            # no image provided for white so use the image for black and
            # rotate it through 180 degrees
            image = image + "B" + extension
            path = os.path.join(pieces_dir, image)

            pb = GdkPixbuf.Pixbuf.new_from_file(path)
            pb = pb.rotate_simple(GdkPixbuf.PixbufRotation.UPSIDEDOWN)
            piece_pixbuf.append(pb)

        return piece_pixbuf, None

    # called from set_board_colours.py when user loads new custom pieces
    def load_custom_pieces(self, path):
        custom_piece_pixbufs, errmsg = self.load_pixbufs(path)
        # if no errors then set custom pixbufs to the new set
        if errmsg is None:
            self.custom_piece_pixbuf = custom_piece_pixbufs
            self.custom_piece_path = path
        return errmsg

    def getpixbuf(self, piece):
        #print(piece.color,piece.piece_type)
        #piece = piece.symbol()
        #if piece==
        # pieces contains the list of possible pieces
        pieces = [
            "None", "p", "n", "b", "r", "k", "q",
                  "P", "N", "B", "R", "K", "Q"]

        try:
            idx = pieces.index(piece)
        except ValueError as ve:
            traceback.print_exc()
            print("error piece not found, piece =", piece)
            print(len(piece))

        if self.pieceset == "jcchess":
            pixbuf = self.jcchess_piece_pixbuf[idx]
        elif self.pieceset == "eastern":
            pixbuf = self.eastern_piece_pixbuf[idx]
        elif self.pieceset == "western":
            pixbuf = self.western_piece_pixbuf[idx]
        elif self.pieceset == "custom":
            try:
                pixbuf = self.custom_piece_pixbuf[idx]
            except TypeError as te:
                print("error loading custom pieces", te)
                pixbuf = self.jcchess_piece_pixbuf[idx]
                self.pieceset = "jcchess"
        else:
            print("invalid pieceset in getpixbuf in pieces.py:", self.pieceset)
            self.pieceset = "jcchess"
            pixbuf = self.jcchess_piece_pixbuf[idx]   # default to jcchess pieces
        return pixbuf

    def get_pieceset(self):
        return self.pieceset

    def set_pieceset(self, pieceset):
        self.pieceset = pieceset

    def get_custom_pieceset_path(self):
        return self.custom_piece_path

    def set_custom_pieceset_path(self, path):
        self.custom_piece_path = path
